Prepare using more unmodified libstdc++ headers

fbrausse opened the pull request:
This PR contains some changes I needed for the next step, which is dropping some of our C++ headers overriding libstdc++'s ones. Most notably, I guess, the *scanf() family of functions is no longer prefixed with `__ESBMC_` by the clang-c frontend, because the actual symbols `::scanf` etc. are expected to exist by libstdc++. Same for the strtol() family, so declarations of the missing ones are added as well. Besides that, some minor fixes and getting rid of warnings.

fbrausse committed:
[c2goto] headers: add preliminary <sched.h> required by libstdc++

fbrausse committed:
[c2goto] headers: include <sched.h> in <pthread.h> required by libstdc++

fbrausse committed:
[c2goto] headers: add missing declarations of strto*() to <stdlib.h> required for libstdc++

fbrausse committed:
[cpp] headers: pull ::strto*() into std namespace expected by libstdc++

fbrausse committed:
[clang-c] assert invariant + check retval of get_struct_union_class()

fbrausse committed:
[python] rm superfluous move() of temporaries

fbrausse committed:
[util] ignore -Wdeprecated-declarations for LLVM includes, llvm-11 has some of those

fbrausse committed:
[solidity] avoid known LLVM build warnings, same as clang-c/c++ frontends

fbrausse committed:
[util] types: move methods() from struct_type to struct_union_type as unions may have them, too

fbrausse committed:
[goto] rm superfluous case (it is handled later)

fbrausse committed:
[goto, symex, clang-c] do not #define *scanf() as __ESBMC-prefixed, the symbols need to exist for libstdc++ compat

fbrausse committed:
[clang-cpp] vft: assert invariant + simplify context access

fbrausse committed:
[clang-c] convert: run get_struct_union_class_methods_decls() when symbol is already recorded

fbrausse committed:
[github] do not auto-commit format changes to my branches

lucasccordeiro reviewed (approved):
Thanks for submitting this PR, @fbrausse. 

Can I ask you to run this PR over the SV-COMP benchmarks before we merge it into the master?

fbrausse commented:
This PR:
```
Statistics:          23805 Files
  correct:           13781
    correct true:     7575
    correct false:    6206
  incorrect:            24
    incorrect true:     13
    incorrect false:    11
  unknown:           10000
  Score:             20764 (max: 38482)
```
Do we have recent master scores?

lucasccordeiro commented:
From 4 days ago:

````
Statistics:          23805 Files
  correct:           13840
    correct true:     7621
    correct false:    6219
  incorrect:            24
    incorrect true:     13
    incorrect false:    11
  unknown:            9941
  Score:             20869 (max: 38482)

GitHub actions: https://github.com/esbmc/esbmc/actions/runs/7849495009
````

fbrausse commented:
> From 4 days ago:

Thanks, I'll take a look at the diff.

lucasccordeiro commented:
@fbrausse: Can I ask you whether you had a chance to check this diff?

fbrausse commented:
The diff shows a lot of timeout noise, *but* also new unknowns for 3 Linux driver tasks, which now fail in the frontend stage with:
```
Converting
ERROR: Conversion of unsupported clang expr: "TypeTraitExpr" to expression
TypeTraitExpr 0x7835f70 </home/rafaelsa/sv-benchmarks/c/ldv-linux-3.4-simple/32_7_cpp_single_drivers-staging-media-dt3155v4l-dt3155v4l.cil.out.i:23100:89, col:162> 'int' __builtin_types_compatible_p

ERROR: CONVERSION ERROR
```
Looking into it.

fbrausse commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -1054,7 +1055,8 @@ bool clang_c_convertert::get_type(const clang::Type &the_type, typet &new_type)
     /* record in context if not already there */
-    get_struct_union_class(rd);
+    if (get_struct_union_class(rd))
+      return true;
```
It's this change that's causing the 3x new unknowns. I don't think it is controversial: checking whether we could actually convert the declaration is not wrong, as opposed to ignoring that result.

It fails for this declaration:
```c

static const struct kparam_array __param_arr_gpio_tab = {
        .max = (sizeof(gpio_tab) / sizeof((gpio_tab)[0]) + (sizeof(struct {
                  int
                  : -!!(__builtin_types_compatible_p(
                            typeof((gpio_tab)), typeof(&(gpio_tab)[0])));
                }))),
        .num = ((void *)0),
        .ops = &param_ops_ushort,
        .elemsize = sizeof(gpio_tab[0]),
        .elem = gpio_tab};
```
specifically the declaration of the struct with the `__builtin_types_compatible_p` as part of the width of the bitfield.

I'll look into adding support for this expression type.

lucasccordeiro replied:
Thanks for checking this, @fbrausse.

fbrausse committed:
[clang-c] handle TypeTraitExpr used for __builtin_types_compatible_p()

This re-enables SV-COMP's

- c/ldv-linux-3.4-simple/32_7_cpp_single_drivers-net-phy-dp83640.cil.out.i
- c/ldv-linux-3.4-simple/32_7_cpp_single_drivers-net-wireless-mwl8k.cil.out.i
- c/ldv-linux-3.4-simple/32_7_cpp_single_drivers-staging-media-dt3155v4l-dt3155v4l.cil.out.i

which did fail with on that builtin, see
<https://github.com/esbmc/esbmc/pull/1674#discussion_r1494402724>.

fbrausse commented:
I've added partial support for the TypeTraitExpr to the clang-c frontend. Needed to import \<clang/ExprCXX.h\> though. Also rebased to master and started [another SV-COMP run](https://github.com/esbmc/esbmc/actions/runs/7959472784) to compare to [master](https://github.com/esbmc/esbmc/actions/runs/7958127390).

fbrausse committed:
[clang-cpp] move TypeTraitExpr handling to clang-c frontend

fbrausse committed:
[regression] fixed KNOWNBUGs

fbrausse commented:
Interesting error on Windows (related to MSVC's C++ lib):
```
Parsing D:/a/esbmc/esbmc/regression/k-induction\cpp_stack_empty_bug\main.cpp
In file included from D:/a/esbmc/esbmc/regression/k-induction\cpp_stack_empty_bug\main.cpp:3:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4\stack:4:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/deque:4:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iterator:4:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iostream:4:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iomanip:4:
In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/definitions.h:4:
In file included from C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\cstddef:8:
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\yvals_core.h:887:1: error: static_assert failed "Error in C++ Standard Library usage."
_EMIT_STL_ERROR(STL1000, "Unexpected compiler version, expected Clang 16.0.0 or newer.");
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\yvals_core.h:514:5: note: expanded from macro '_EMIT_STL_ERROR'
    static_assert(false, "Error in C++ Standard Library usage.")
    ^             ~~~~~
```
I'll keep the Windows-version as known-bug.

mikhailramalho commented:
we probably need to update minimum clang version?

Em seg., 19 de fev. de 2024 às 13:24, Franz Brauße ***@***.***>
escreveu:

> Interesting error on Windows (related to MSVC's C++ lib):
>
> Parsing D:/a/esbmc/esbmc/regression/k-induction\cpp_stack_empty_bug\main.cpp
> In file included from D:/a/esbmc/esbmc/regression/k-induction\cpp_stack_empty_bug\main.cpp:3:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4\stack:4:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/deque:4:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iterator:4:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iostream:4:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/iomanip:4:
> In file included from C:\Users\RUNNER~1\AppData\Local\Temp\esbmc-cpp-headers-83f8-544f-99c4/definitions.h:4:
> In file included from C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\cstddef:8:
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\yvals_core.h:887:1: error: static_assert failed "Error in C++ Standard Library usage."
> _EMIT_STL_ERROR(STL1000, "Unexpected compiler version, expected Clang 16.0.0 or newer.");
> ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.38.33130\include\yvals_core.h:514:5: note: expanded from macro '_EMIT_STL_ERROR'
>     static_assert(false, "Error in C++ Standard Library usage.")
>     ^             ~~~~~
>
> I'll keep the Windows-version as known-bug.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1674#issuecomment-1952817202>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6A2FCM2NLHUKJPMK3YUN4EBAVCNFSM6AAAAABDGX4WB2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNJSHAYTOMRQGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

fbrausse committed:
[regression] cmake: add windows/ subdirectory

fbrausse committed:
[regression] dup buggy k-induction/cpp_stack_empty_bug

See <https://github.com/esbmc/esbmc/pull/1674#issuecomment-1952817202>.

fbrausse commented:
> we probably need to update minimum clang version?

I guess so, at least to use the MSVC++ libs on Windows. For now it should be "fixed" by having a separate KNOWNBUG copy for Windows only.

fbrausse commented:
Regarding Clang-16, we still haven't addressed this issue: #862

fbrausse commented:
Master (same basis as #1689):
```
Statistics:          23805 Files
  correct:           13826
    correct true:     7608
    correct false:    6218
  incorrect:            25
    incorrect true:     13
    incorrect false:    12
  unknown:            9954
  Score:             20826 (max: 38482)

https://github.com/esbmc/esbmc/actions/runs/7958127390
```
This PR:
```
Statistics:          23805 Files
  correct:           13845
    correct true:     7626
    correct false:    6219
  incorrect:            24
    incorrect true:     13
    incorrect false:    11
  unknown:            9936
  Score:             20879 (max: 38482)

https://github.com/esbmc/esbmc/actions/runs/7959472784
```

lucasccordeiro commented:
Thanks for submitting this PR, @fbrausse.
